core/src/main/scala/akka/persistence/r2dbc/internal/h2/H2Dialect.scala

```diff
@@ -0,0 +1,165 @@
+    val res = settings
+      // app timestamp is db timestamp because same process
+      .withUseAppTimestamp(true)
+      .withDbTimestampMonotonicIncreasing(true)
+    res
+  }
+
+  override def createConnectionFactory(settings: R2dbcSettings, config: Config): ConnectionFactory = {
+    def r2option[T](h2Option: H2ConnectionOption): io.r2dbc.spi.Option[T] =
+      io.r2dbc.spi.Option.valueOf[T](h2Option.getKey)
+
+    // starting point for both url and regular configs,
+    // to allow url to override anything but provide sane defaults
+    val builder = ConnectionFactoryOptions.builder()
+    val createSliceIndexes = config.getBoolean("create-slice-indexes")
+    builder
+      .option(ConnectionFactoryOptions.DRIVER, "h2")
+      // create schema on first connect
+      .option(
+        r2option(H2ConnectionOption.INIT),
+        dbSchema(settings, createSliceIndexes) + config.getString("additional-init"))
+      // don't auto close connections
+      .option(r2option(H2ConnectionOption.DB_CLOSE_DELAY), "-1")
+
+    if (config.getBoolean("trace-logging"))
+      // log to SLF4J instead of print to stdout, logger name will be 'h2database'
+      builder.option(r2option(H2ConnectionOption.TRACE_LEVEL_FILE), "4")
+
+    val url = config.getString("url")
+    if (url.isEmpty) {
+      builder
+        // note: protocol other than mem or file is validated in r2db driver
+        .option(ConnectionFactoryOptions.PROTOCOL, config.getString("protocol"))
+        .option(ConnectionFactoryOptions.DATABASE, config.getString("database"))
+    } else {
+      val urlConfig = ConnectionFactoryOptions
+        .parse(url)
+        .mutate()
+        .build()
+      builder.from(urlConfig)
+    }
+
+    ConnectionFactories.get(builder.build())
+  }
+
+  override def createJournalDao(settings: R2dbcSettings, connectionFactory: ConnectionFactory)(implicit
+      system: ActorSystem[_]): JournalDao = new H2JournalDao(settings, connectionFactory)(ecForDaos(system), system)
+
+  override def createSnapshotDao(settings: R2dbcSettings, connectionFactory: ConnectionFactory)(implicit
+      system: ActorSystem[_]): SnapshotDao = new H2SnapshotDao(settings, connectionFactory)(ecForDaos(system), system)
+
+  override def createQueryDao(settings: R2dbcSettings, connectionFactory: ConnectionFactory)(implicit
+      system: ActorSystem[_]): QueryDao = new H2QueryDao(settings, connectionFactory)(ecForDaos(system), system)
+
+  override def createDurableStateDao(settings: R2dbcSettings, connectionFactory: ConnectionFactory)(implicit
+      system: ActorSystem[_]): DurableStateDao =
+    new H2DurableStateDao(settings, connectionFactory)(ecForDaos(system), system)
+
+  private def ecForDaos(system: ActorSystem[_]): ExecutionContext = {
+    // H2 R2DBC driver blocks in surprising places (Mono.toFuture in stmt.execute().asFuture())
+    // FIXME is the default blocking good enough, should we have a separate to compartmentalize?
```

Comment:
When I profiled it looked like it would always be blocking for multithreaded access, but perhaps the time an operation to complete with in-mem is so short that it doesn't matter.